diffrax/_solver/langevin_srk.py

```diff
@@ -0,0 +1,289 @@
+        # else:
+        #     c_axes = tuple(range(c.ndim))
+        #     out = jax.vmap(self._tay_cfs_single, in_axes=c_axes, out_axes=c_axes)(c)
+        out = self._tay_cfs_single(c)
+
+        def check_shape(coeff_leaf):
+            permitted_shapes = [c.shape + (3, 6), c.shape + (1, 6), c.shape + (6,)]
+            assert (
+                coeff_leaf.shape in permitted_shapes
+            ), f"leaf shape: {coeff_leaf.shape}, c shape: {c.shape}"
+
+        jtu.tree_map(check_shape, out)
+        return out
+
+    @staticmethod
+    def _eval_taylor(h, tay_cfs: _Coeffs) -> _Coeffs:
+        # Multiplies the pre-computed Taylor coefficients by powers of h.
+        # jax.debug.print("eval taylor for h = {h}", h=h)
+        dtype = tay_cfs.dtype
+        h_powers = jnp.power(h, jnp.arange(0, 6, dtype=h.dtype)).astype(dtype)
+        return jtu.tree_map(
+            lambda tay_leaf: jnp.tensordot(tay_leaf, h_powers, axes=1), tay_cfs
+        )
+
+    def _recompute_coeffs(
+        self, h, gamma: LangevinX, tay_cfs: PyTree[_Coeffs]
+    ) -> _Coeffs:
+        def recompute_coeffs_leaf(c: ArrayLike, _tay_cfs: _Coeffs):
+            # Used when the step-size h changes and coefficients need to be recomputed
+            # Depending on the size of h*gamma choose whether the Taylor expansion or
+            # direct computation is more accurate.
+            cond = h * c < self.taylor_threshold  # c is a leaf of gamma
+            if jnp.ndim(c) == 0:
+                return lax.cond(
+                    cond,
+                    lambda h_: self._eval_taylor(h_, _tay_cfs),
+                    lambda h_: self._directly_compute_coeffs_leaf(h_, c),
+                    h,
+                )
+            else:
+                tay_out = self._eval_taylor(h, _tay_cfs)
+                if cond.ndim < jtu.tree_leaves(tay_out)[0].ndim:
+                    cond = jnp.expand_dims(cond, axis=-1)
+
+                def select_tay_or_direct(dummy):
+                    fun = lambda _c: self._directly_compute_coeffs_leaf(h, _c)
+                    direct_out = vmap(fun)(c)
+
+                    def _choose(tay_leaf, direct_leaf):
+                        assert tay_leaf.ndim == direct_leaf.ndim == cond.ndim, (
+                            f"tay_leaf.ndim: {tay_leaf.ndim},"
+                            f" direct_leaf.ndim: {direct_leaf.ndim},"
+                            f" cond.ndim: {cond.ndim}"
+                        )
+                        return jnp.where(cond, tay_leaf, direct_leaf)
+
+                    return jtu.tree_map(_choose, tay_out, direct_out)
+
+                # If all entries of h*gamma are below threshold, only compute tay_out
+                # otherwise, compute both tay_out and direct_out and select the
+                # correct one for each dimension
+                return lax.cond(
+                    jnp.all(cond), lambda _: tay_out, select_tay_or_direct, None
+                )
```

Comment:
Do you have any opinion on how this should behave around batching?

If the predicate of a `lax.cond` is batched then both branches are computed unconditionally and then `jnp.where`'d together.

If you're doing this logic here just for efficiency, then you might want to swap out the predicate for something like `eqxi.unvmap_all`, whilst taking care to ensure that the internal logic of both branches remains valid.

Reply:
Yes, um I agree this should be improved. I'll keep it pending for now and address it in alater commit.

Reply:
I thought this would require more work, but yeah simply substituting in `eqxi.unvmap_all` worked like a charm. And indeed both branches are always logically valid, that's for sure. This is a very useful trick, thanks for recommending!